### PR TITLE
Implement ProductBuilder pattern throughout tests

### PR DESCRIPTION
**Pull request by andrewHEguardian**
## What does this change?

This PR is for updating tests to use the productBuilder pattern (#1069 )
- Replace all `productDetail.ts` test objects with `testProducts.ts` functions
- Update all references in tests
- Where necessary, tests changed to be less hardcoded to the test subscription object

## How to test

Cypress, Chromatic and Jest tests all pass with as few changes as possible. 

## How can we measure success?

Fewer LoC to add a new test subscription variation

## Accessibility

-   [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-   [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-   [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

### CODE REVIEW

File: cypress/integration/parallel-4/deliveryRecords.spec.ts [hunk 1]
```diff
@@ -16,43 +12,47 @@ import {
...
 		signInAndAcceptCookies();
 
 		cy.intercept('GET', '/api/me/mma?productType=Weekly', {
 			statusCode: 200,
-			body: toMembersDataApiResponse(guardianWeeklyCurrentSubscription),
+			body: toMembersDataApiResponse(guardianWeeklyPaidByCard()),
 		}).as('product_detail');
 
-		cy.intercept('GET', '/api/delivery-records/A-S00293857', {
+		cy.intercept('GET', '/api/delivery-records/*', {
```
Comment (charleycampbell) [inline cypress/integration/parallel-4/deliveryRecords.spec.ts:29]: Why are we not passing in the sub id now?

**Review by charleycampbell — APPROVED** [review]
Looks great! Only tiny comments/clarifications 👍

File: client/fixtures/productBuilder/baseProducts.ts [hunk 2]
```diff
@@ -135,16 +163,271 @@ export function baseDigitalPack(): ProductDetail {
...
+		isPaidTier: true,
+		selfServiceCancellation: {
+			isAllowed: true,
+			shouldDisplayEmail: true,
+			phoneRegionsToDisplay: ['UK & ROW', 'US', 'AUS'],
+		},
+		joinDate: '2022-01-05',
+		optIn: true,
+		subscription: {
+			contactId: '0039E00001KA26BQAT',
```
[the remark this answers is not in the document]
Comment (andrewHEguardian) [reply client/fixtures/productBuilder/baseProducts.ts:209]: yes, from the old productDetails. Could make them all more obviously test data

File: cypress/integration/parallel-4/deliveryRecords.spec.ts (hunk 1 above)
Comment (andrewHEguardian) [reply cypress/integration/parallel-4/deliveryRecords.spec.ts:29]: the sub ID changed, and it seemed fragile to have it be hardcoded in the url when we can wildcard it away. If you think the logic of the test means we should only intercept that specific sub maybe we could specify it from the test

File: client/fixtures/productBuilder/baseProducts.ts [hunk 3]
```diff
@@ -15,32 +15,60 @@ import type { ProductDetail } from '../../../shared/productResponse';
...
-			nextPaymentPrice: null,
-			nextPaymentDate: null,
-			lastPaymentDate: null,
-			chargedThroughDate: null,
-			renewalDate: '2022-12-16',
-			anniversaryDate: '2022-12-16',
+			start: '2023-04-26',
+			end: '2024-04-26',
+			nextPaymentPrice: 700,
+			nextPaymentDate: '2023-05-26',
```
[the remark this answers is not in the document]
Comment (charleycampbell) [reply client/fixtures/productBuilder/baseProducts.ts:34]: Makes sense!

File: client/fixtures/productBuilder/baseProducts.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (charleycampbell) [reply client/fixtures/productBuilder/baseProducts.ts:209]: If we know they're test I think it's fine...although if we can make obviously test data might save future questions!

File: cypress/integration/parallel-4/deliveryRecords.spec.ts (hunk 1 above)
Comment (charleycampbell) [reply cypress/integration/parallel-4/deliveryRecords.spec.ts:29]: good point about it being unstable...I do wonder though whether we should make it obvious that it's the correct sub id but from the looks of it we're only dealing with one sub id in the test anyway

Comment (andrewHEguardian) [reply cypress/integration/parallel-4/deliveryRecords.spec.ts:29]: there is another test that has two subs, but the logic is basically around going directly to the page or not and still works

File: client/fixtures/productBuilder/baseProducts.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewHEguardian) [reply client/fixtures/productBuilder/baseProducts.ts:209]: One for a future PR, luckily now there are fewer places to change!

**Pull request merged**